### PR TITLE
Add new hierachical clustering tool based on scipy

### PR DESCRIPTION
**Pull request by wm75**
A tool like this was requested by @dbrites

FOR CONTRIBUTOR:
* [X] I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] License permits unrestricted use (educational + commercial)
* [X] This PR adds a new tool or tool collection
* [ ] This PR updates an existing tool or tool collection
* [ ] This PR does something else (explain below)

### CODE REVIEW

**Comment by wm75** [comment]
One thing this version does not support atm is *multiple* heights or n_clusters values, which, on the command line would result in multi-column tabular output with the cluster assignments for all different settings.
Let me know if that seems of interest.

File: tools/clustering_from_distmat/clustering_from_distmat.xml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                <option value="height">Yes, and use distance threshold to divide into clusters</option>
+            </param>
+            <when value="dendrogram-only" />
+            <when value="n-cluster">
+                <param name="n_cluster" type="integer" value="5" min="1" label="How many clusters to divide into?" />
+                <param name="generate_dendrogram" type="boolean" label="Produce also the dendrogram of clustering results" />
+            </when>
+            <when value="height">
+                <param name="height" type="float" value="5.0" label="Distance threshold for clusters to be reported" />
+                <param name="generate_dendrogram" type="boolean" label="Produce also the dendrogram of clustering results" />
```
Comment (bgruening) [inline tools/clustering_from_distmat/clustering_from_distmat.xml:42]: This param is not used in the CLI

Comment (wm75) [reply tools/clustering_from_distmat/clustering_from_distmat.xml:49]: Yes, this is difficult (and not so important) to implement in the cli.
Instead this is used only to filter the outputs on the Galaxy side to avoid clutter in the history.
Not a good idea?

File: tools/clustering_from_distmat/.shed.yml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+name: clustering_from_distmat
+owner: iuc
+description: Distance matrix-based hierarchical clustering using SciPy
+remote_repository_url: https://github.com/galaxyproject/tools-iuc/tree/master/tools/clustering_from_distmat/
```
Comment (bgruening) [inline tools/clustering_from_distmat/.shed.yml:4]:
```suggestion
remote_repository_url: https://github.com/galaxyproject/tools-iuc/tree/master/tools/clustering_from_distmat/
long_description: Distance matrix-based hierarchical clustering using SciPy
homepage_url: https://github.com/galaxyproject/tools-iuc/tree/master/tools/clustering_from_distmat/
```

**Review by bgruening — APPROVED** [review]

**Comment by wm75** [comment]
We should maybe have column names for the cluster assignment output. Let me add this still before merging.

File: tools/clustering_from_distmat/test-data/test_assignment_average_h18.tsv [hunk 3]
```diff
@@ -0,0 +1,6 @@
+sample	cluster_id_h18.0
```
Comment (bgruening) [inline tools/clustering_from_distmat/test-data/test_assignment_average_h18.tsv:1]: why is this here h18.0 a float and below an integer? Is anything wired going on?
Any both files are identical besides the header...

Comment (wm75) [reply tools/clustering_from_distmat/test-data/test_assignment_average_h18.tsv:1]: h18.0 is the height at which the tree got cut in order to produce the cluster IDs below, and height can be a float.
The results are otherwise identical because I chose --height and -n so that both tests give the same result. So before I added now the column names only a single test file was needed.
